File: 1370-count-number-of-nice-subarrays/1370-count-number-of-nice-subarrays.java
class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
        return nice(nums,k)-nice(nums,k-1);
    }

    private int nice(int[]arr,int k)
    {
        int r=0,l=0,sum=0,result=0;

        int n=arr.length;

        while(r<n)
        {
            if(arr[r]%2!=0)
            {
                sum=sum+1;
            }

            while(sum>k )
            {
                sum=sum-arr[l]%2;
                l++;
            }
            result=result+(r-l+1);
            r++;

        }

        return result;
    }
}

File: 1370-count-number-of-nice-subarrays/1370-count-number-of-nice-subarrays.py
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
        
        nums = [1 if num % 2 == 1 else 0 for num in nums]
        
        def atmost(nums, k) -> int:
            count = 0
            left = 0
            sum_ = 0

            for right in range(len(nums)):
                k -= nums[right]
                while k < 0:
                    k += nums[left]
                    left += 1
                count += (right - left + 1)
            return count

        return atmost(nums, k) - atmost(nums, k - 1)
